<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Nice)</span></div>

## 1 There is more…

### 1.1 There is more to list comprehension

In [1]:
# Your code here
# example 1
[[a,b] for a in range(5) for b in ['A', 'B', 'C']]

[[0, 'A'],
 [0, 'B'],
 [0, 'C'],
 [1, 'A'],
 [1, 'B'],
 [1, 'C'],
 [2, 'A'],
 [2, 'B'],
 [2, 'C'],
 [3, 'A'],
 [3, 'B'],
 [3, 'C'],
 [4, 'A'],
 [4, 'B'],
 [4, 'C']]

In [2]:
# example 2
# incorporating condition
[[a,b] for a in range(5) for b in ['A', 'B', 'C'] if a % 2 != 0]

[[1, 'A'], [1, 'B'], [1, 'C'], [3, 'A'], [3, 'B'], [3, 'C']]

In [3]:
# example 3
# flatten a list
nested_list=[[1, 2, 3], [4, 5, 6, 7]]
[y for x in nested_list for y in x]

[1, 2, 3, 4, 5, 6, 7]

### 1.2 Zipping a dictionary

In [4]:
# Your code here
super_names=["Black Widow", "Iron Man", "Doctor Strange"]
real_names=["Natasha Romanoff", "Tony Stark", "Stephen Strange"]

dict(zip(real_names, super_names))

# dict() is used to recast zip()'s output into a dictionary

{'Natasha Romanoff': 'Black Widow',
 'Tony Stark': 'Iron Man',
 'Stephen Strange': 'Doctor Strange'}

### 1.3 for and while has an else

In [5]:
# Your code here
#'for' and 'while' loops come with optional 'else'
#'else' executed only if loops are completed
#'else' won't run if exit prematurely(by using 'break')

numbers=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

for i in numbers:
    if i < 0: break
else:
    print('No negative numbers in the list')

No negative numbers in the list


In [6]:
# adding negative number = no message
# as there is a break and 'else' doesn't run

numbers=[-1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

for i in numbers:
    if i < 0: break
else:
    print('No negative numbers in the list')

## Explore 1 :  Changing a list

In [7]:
fruits = ["apple", "banana", "jackfruit",
          "pineapple", "papaya", "watermelons",
          "peaches", "durian",  "mangoes",
          "strawberries", "passionfruit"
          ]

In [16]:
# remove fruits starting with 'p'

for fruit in fruits:
    if fruit[0:10] == "p":
        fruits.remove(fruit)
print(fruits)

['apple', 'banana', 'jackfruit', 'watermelons', 'durian', 'mangoes', 'strawberries']


## Explore 2 :  A list of power

In [9]:
# attempt to create a list [n, n^2, n^3] for several values of n
# specify max value by changing 'maximum_n'

maximum_n = 5
result = [[] for i in range(5)]

for n in range(1, maximum_n + 1):  # maximum_n + 1 when run = maximum_n
    result[n - 1].append(n)
    result[n - 1].append(n**2) 
    result[n - 1].append(n**3) 
print(result)

[[1, 1, 1], [2, 4, 8], [3, 9, 27], [4, 16, 64], [5, 25, 125]]


for `maximum_n = 5`, content of `result` should be:
[[1, 1, 1],
 [2, 4, 8],
 [3, 9, 27],
 [4, 16, 64],
 [5, 25, 125]]

## Explore 3 :  Time profiling

Use `%%timeit` to compare the execution speeds of the following:

|#|**Option 1**|**Option 2**|**Result**|
|:---|:--:|:--:|---:|
|1|Creating a list of squares with `for` loop|Creating a list of squares with `while` loop
|2|Creating a list of squares with a `for` loop|Creating a list of squares with a **list comprehension** loop
|3|Creating a list of squares using **list** `append()`|Creating a list of squares using **list** `+=`
|4|Creating a list of squares using **list** `append()`|Creating a list of squares using `append()` of **Numpy**
|5|Creating a list of squares using **Numpy**|Creating a list of squares using **List comprehension** loop

Kiwi: Lowkey very tedious ;-; will try to get back when I actually have time